### ISSUE TITLE
win_robocopy: Causing changes in win_copy or win_template tasks

### ISSUE THREAD

**Issue by dagwieers** [ansible/ansible#25302]
_From @davidhulick on September 1, 2016 19:0_

##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

win_template
##### ANSIBLE VERSION

```
ansible 2.1.1.0
```
##### OS / ENVIRONMENT

Running Ansible from CentOS Linux release 7.2.1511 (Core)
Running Ansible against Windows Server 2012R2
##### SUMMARY

win_template always reports changed, even when only variable is consistent between runs.
##### STEPS TO REPRODUCE

```
task in playbook:
- name: Wolfpack name template.
  win_template:
        src: "templates/role.castle.config.j2"
        dest: "C:\\Program Files (x86)\\Wolfpack\\config\\role.castle.config"

template: https://gist.github.com/davidhulick/487d5417f8839634fe8539ea350ed36b

Relevant line in inventory file, where variable is defined: 
iis01 ansible_host=192.168.1.105 wolfpack_name="IIS 01"

```

##### EXPECTED RESULTS

Task reported as green OK in playbook output.
##### ACTUAL RESULTS

```
https://gist.github.com/davidhulick/932d85cd286299a02a23b060edb9cb92
```

_Copied from original issue: ansible/ansible-modules-core#4636_

**Comment by dagwieers** [comment]
_From @ansibot on September 1, 2016 19:3_



@jhawkesworth ping, this issue is waiting for your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dagwieers** [comment]
_From @jhawkesworth on September 1, 2016 22:12_

Last time I looked, the win_ template action was a very thin wrapper around the template action. So I am curious if template behaves the same way (on Linux hosts). If template on Linux is behaving as you expect, then I suspect the issue would be in win_copy, win_file or possibly in the shared module code in powershell.ps1 

I will investigate as soon as I get a chance.

**Comment by dagwieers** [comment]
_From @davidhulick on September 1, 2016 23:35_

template running against a CentOS 6.8 host works as intended.

**Comment by dagwieers** [comment]
_From @jhawkesworth on September 15, 2016 16:48_

Hi @davidhulick Sorry it has taken me a while to get to this.

I just tried against latest devel and not reproduced using the following playbook

```
jon@TENSY:~$ cat 4636-test.yml
---

- hosts: 10t
  gather_facts: false
  tasks:
   - name: test templating a file that shouldn't change
     win_template:
        src: test.j2
        dest: test.txt

```

and this template

```
jon@TENSY:~$ cat test.j2
This file has nothing in it that should change.
There is just one var here {{ test_variable }}
That is all
```

 Command line used was
`ansible-playbook -v 4636-test.yml -e "test_variable=the_same"
`
How are you setting your variable that isn't changing?  inventory? group_vars? role defaults?  I guess its possible that it looks like it has changed in someway.

needs_info

**Comment by dagwieers** [comment]
_From @ansibot on October 5, 2016 15:9_

@davidhulick ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dagwieers** [comment]
_From @ansibot on October 21, 2016 8:36_

@davidhulick ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dagwieers** [comment]
_From @ansibot on November 6, 2016 9:35_

@davidhulick ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dagwieers** [comment]
_From @ansibot on November 22, 2016 9:43_

@davidhulick ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dagwieers** [comment]
_From @ansibot on December 7, 2016 18:12_

This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by dagwieers** [comment]
_From @davidhulick on January 5, 2017 15:30_

Hi @jhawkesworth Sorry for the very late reply on this.

Here is where the variable is being called in my template file role.castle.config.j2

```
<parameters>
   <SiteId>{{ wolfpack_name }}</SiteId>
   <Longitude></Longitude>
   <Latitude></Latitude>
</parameters>
```

This variable is defined in the inventory file, and is different for each machine I am managing.

```
vmaqcmxtdev13 ansible_host=10.63.61.49 wolfpack_name="VMA QC MXT DEV13" nagios_hostgroup="qcservers"
vmaqcmxtdev14 ansible_host=10.63.61.50 wolfpack_name="VMA QC MXT DEV14" nagios_hostgroup="qcservers"
vmaqcmxtdev15 ansible_host=10.63.61.51 wolfpack_name="VMA QC MXT DEV15" nagios_hostgroup="qcservers"
```

I have run a checksum before and after each ansible run, and the checksum on that template file (after its copied to windows) remains the same. The modified timestamp does not change unless I run ansible.

**Comment by dagwieers** [comment]
_From @jhawkesworth on January 5, 2017 16:59_

I just tried using an inventory var against latest devel against a win10 and a s2012r2 vm and I still can't reproduce this problem. :-(

S2012R2 is lazy about updating modified timestamps in my experience (I think it only updates on a file close, unlike previous versions of windows).  
Can you use resource monitor to check if anything else is holding a handle to the file on your windows boxes?  Also, what happens if you stop IIS and then run your playbook twice.  Does it still report changed on the second run?

needs_info

**Comment by dagwieers** [comment]
_From @davidhulick on January 18, 2017 19:46_

Hi @jhawkesworth,

It looks like I've finally gotten to the bottom of this. It turns out the issue was the win_robocopy ansible module. During initial deployment, I used the following code to copy our Wolfpack directory to Program Files (x86)

````
- name: Copy wolfpack.
  win_robocopy:
        src: "C:\\ansible_setup_applications\\Wolfpack"
        dest: "C:\\Program Files (x86)\\Wolfpack"
        recurse: true
````

It turns out that on subsequent runs, even though this step was reporting changed: false, it was affecting the template files in the directory below. When I removed this robocopy step, the templates would compare and report as unchanged as expected.

My temporary workaround is to zip the initial wolfpack "seed" folder and use the following step instead

````
- name: Copy wolfpack.
  win_unzip:
        src: "C:\\ansible_setup_applications\\Wolfpack.zip"
        dest: "C:\\Program Files (x86)"
        creates: "C:\\Program Files (X86)\\Wolfpack"
````

This works as intended. Do you think this is worth opening a new bug for the win_robocopy module? 

Thanks for your help!

**Comment by dagwieers** [comment]
@davidhulick If this issue is still in v2.3 or the devel branch, please open a bug in https://github.com/ansible/ansible ! A module shouldn't change anything when it reports `changed: no`.

**Comment by dagwieers** [comment]
At [19:00 UTC](http://www.thetimezoneconverter.com/?t=19:00&tz=UTC) this ticket will be processed during [the first Windows Sprint](https://github.com/ansible/community/wiki/Windows:-sprints#2017-07-04-1900-utc-sprint-1), we would appreciate if you could join the Sprint to help out with fixing/closing this issue.

**Comment by jhawkesworth** [comment]
Discussed at the sprint.  We believe that win_robocopy previously did not report changed: false reliably as certain return codes were not handled.  Once https://github.com/ansible/ansible/pull/25410 has been merged this should be re-tested.

**Comment by dagwieers** [comment]
@davidhulick Can you please retest your issue with the latest devel branch ? It now has some win_robocopy fixes integrated which we think may fix your issue.

**Comment by dagwieers** [comment]
Since we haven't heard back from the original submitter, and there's a good chance this is fixed in the v2.4 release, I am going to consider this one closed. Feel free to reopen if this is still causing issues.
